consensus/src/ordered_broadcast/types.rs

```diff
@@ -0,0 +1,497 @@
+        let digest = D::read(reader)?;
+        let epoch = Epoch::read(reader)?;
+        let signature = Signature::read(reader)?;
+        Ok(Self {
+            digest,
+            epoch,
+            signature,
+        })
+    }
+}
+
+impl<D: Digest> FixedSize for Parent<D> {
+    const SIZE: usize = D::SIZE + Epoch::SIZE + Signature::SIZE;
+}
+
+/// Node is a message from a sequencer that contains a Chunk and a proof that the parent was correctly broadcasted.
+///
+/// It represents a newly-proposed tip of the chain for the given sequencer.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Node<C: Verifier, D: Digest> {
+    /// Chunk of the node.
+    pub chunk: Chunk<C::PublicKey, D>,
+
+    /// Signature of the sequencer the chunk.
+    pub signature: C::Signature,
+
+    /// Information about the parent chunk
+    ///
+    /// This part is not signed over, but it is used to verify that the previous chunk in the chain was correctly broadcast.
+    pub parent: Option<Parent<D>>,
+}
+
+impl<C: Verifier, D: Digest> Node<C, D> {
+    /// Create a new node with the given chunk, signature, and parent.
+    pub fn new(
+        chunk: Chunk<C::PublicKey, D>,
+        signature: C::Signature,
+        parent: Option<Parent<D>>,
+    ) -> Self {
+        Self {
+            chunk,
+            signature,
+            parent,
+        }
+    }
+
+    pub fn verify(
+        &self,
+        public: Option<&Public>,
+        chunk_namespace: &[u8],
+        ack_namespace: &[u8],
+    ) -> Result<Option<Chunk<C::PublicKey, D>>, Error> {
+        // Verify chunk
+        let message = self.chunk.encode();
+        if !C::verify(
+            Some(chunk_namespace),
+            &message,
+            &self.chunk.sequencer,
+            &self.signature,
+        ) {
+            return Err(Error::InvalidSignature);
+        }
+        let Some(parent) = &self.parent else {
+            return Ok(None);
+        };
+
+        // Verify parent (if present)
+        let Some(public) = public else {
+            unreachable!("public should always be present when parent is present");
+        };
+        let parent_chunk = Chunk::new(
+            self.chunk.sequencer.clone(),
+            self.chunk.height - 1, // Will not parse if height is 0 and parent exists
+            parent.digest,
+        );
+        if !verify_lock(
+            public,
+            &parent_chunk,
+            &parent.epoch,
+            &parent.signature,
+            ack_namespace,
+        ) {
+            return Err(Error::InvalidSignature);
+        }
+        Ok(Some(parent_chunk))
+    }
+}
+
+impl<C: Verifier, D: Digest> Write for Node<C, D> {
+    fn write(&self, writer: &mut impl BufMut) {
+        self.chunk.write(writer);
+        self.signature.write(writer);
+        if let Some(parent) = &self.parent {
+            true.write(writer);
+            parent.write(writer);
+        } else {
+            false.write(writer);
+        }
```

Comment:
`Option<T>` is implemented as this already in `codec`. Just do `self.parent.write(buf)` and it will do this for you :)

Reply:
I vaguely recalled this being implemented. Very nice.